src/encoder.rs

```diff
@@ -24,140 +21,150 @@ use byteorder::{BigEndian, ByteOrder};
 /// );
 /// assert!(encoder::encode(&packet).is_ok())
 /// ```
-pub fn encode(packet: &OscPacket) -> Result<Vec<u8>> {
+pub fn encode(packet: &OscPacket) -> crate::types::Result<Vec<u8>> {
+    let mut bytes = Vec::new();
+    encode_into(packet, &mut bytes)
+        .expect("Failed to write encoded packet into Vec");
+    Ok(bytes)
+}
+
+/// Takes a reference to an OSC packet and writes the
+/// encoded bytes to the given output. On success, the
+/// number of bytes written will be returned. If an error
+/// occurs during encoding, encoding will stop and the
+/// error will be returned. Note that in that case, the
+/// output may have been partially written!
+///
+/// NOTE: The OSC encoder will write output in small pieces
+/// (as small as a single byte), so the output should be
+/// buffered if write calls have a large overhead (e.g.
+/// writing to a file or a socket).
+///
+/// # Example
+///
+/// ```
+/// use rosc::{OscPacket,OscMessage,OscType};
+/// use rosc::encoder;
+///
+/// let mut bytes = Vec::new();
+/// let packet = OscPacket::Message(OscMessage{
+///         addr: "/greet/me".to_string(),
+///         args: vec![OscType::String("hi!".to_string())]
+///     }
+/// );
+/// assert!(encoder::encode_into(&packet, &mut bytes).is_ok())
+/// ```
+pub fn encode_into<O: Output>(packet: &OscPacket, out: &mut O) -> Result<usize, O::Err> {
     match *packet {
-        OscPacket::Message(ref msg) => encode_message(msg),
-        OscPacket::Bundle(ref bundle) => encode_bundle(bundle),
+        OscPacket::Message(ref msg) => encode_message(msg, out),
+        OscPacket::Bundle(ref bundle) => encode_bundle(bundle, out),
     }
 }
 
-fn encode_message(msg: &OscMessage) -> Result<Vec<u8>> {
-    let mut msg_bytes: Vec<u8> = Vec::new();
-
-    msg_bytes.extend(encode_string(msg.addr.clone()));
-    let mut type_tags: Vec<char> = vec![','];
-    let mut arg_bytes: Vec<u8> = Vec::new();
+fn encode_message<O: Output>(msg: &OscMessage, out: &mut O) -> Result<usize, O::Err> {
+    let mut written = encode_string_into(&msg.addr, out)?;
 
+    written += out.write(b",")?;
     for arg in &msg.args {
-        let (bytes, tags): (Option<Vec<u8>>, String) = encode_arg(arg)?;
-
-        type_tags.extend(tags.chars());
-        if let Some(data) = bytes {
-            arg_bytes.extend(data);
-        }
+        written += encode_arg_type(arg, out)?;
     }
 
-    msg_bytes.extend(encode_string(type_tags.into_iter().collect::<String>()));
-    if !arg_bytes.is_empty() {
-        msg_bytes.extend(arg_bytes);
-    }
-    Ok(msg_bytes)
-}
-
-fn encode_bundle(bundle: &OscBundle) -> Result<Vec<u8>> {
-    let mut bundle_bytes: Vec<u8> = Vec::new();
-    bundle_bytes.extend(encode_string("#bundle".to_string()).into_iter());
+    let padding = pad(written as u64 + 1) as usize - written;
+    written += out.write(&[0u8; 4][..padding])?;
```

Comment:
I think three bytes should be enough, because if it's 4 then the bytes are already padded.

```suggestion
    written += out.write(&[0u8; 3][..padding])?;
```